### ISSUE TITLE
250814 Deep Learning 기본 모델 구조_이가은

### ISSUE THREAD

**Issue by kkaeunii** [AIBootcamp14/wiki-rider#28]
## 📅 학습 날짜  
*2025-08-14*

## 📌 주요 개념
**키워드** : CNN, RNN, ChatGPT

**요약**      
-  CNN (Convolutional Neural Network, 합성곱 신경망) : 딥러닝 모델 중 하나로 이미지 분석에 사용
- RNN (Recurrent Neural Networks, 순환 신경망) : 과거의 정보를 기억하면서 고정되지 않은 현재의 입력도 동시에 처리
- ChatGPT : OpenAI 사의 대화형 인공지능 서비

### CNN (Convolutional Neural Network, 합성곱 신경망)

합성곱 연산으로 모델 구조를 정해 놓아, 이미지 형상 정보를 유지
형상 정보 : 공간적 정보를 의미

- **키워드** : 합성곱 연산, 필터/커널, 피처맵, 스트라이드(Stride), 패딩(Padding), 풀링 연산

- **개념**
   - 합성곱 연산 : 입력 데이터에 필터/커널을 적용해 윈도우를 일정 간격(스트라이드)으로 이동해가며 수행
   - 스트라이드(Stride) : 필터가 이동하는 간격으로 피처맵의 크기를 조절할 때 사용
   - 패딩(Padding) : 합성곱 연산 전, 입력 데이트 주변을 특정 값으로 채우는 연산으로 입력 데이터를 키움 
   - 피처맵 : 연산의 출력 결과         
   - 풀링 연산 : 데이터의 가로, 세로 방향의 크기를 줄이는 연산으로 피처맵에서 중요 정보를 추출하기 위해 사용           
      - 최대 풀링 방식 : 피처맵에서 최대값을 취하는 방식, 중요한 정보만 선별해내는 기능          
      - 평균 풀링 방식 : 피처맵의 평균값을 계산하는 방식, 전반적인 특징을 중요하게 생각할 때 활용                  

### RNN (Recurrent Neural Network, 순환 신경망)

이전 시점의 정보를 현재 시점의 입력과 함께 처리하는 순환 구조로 순차적 데이터나 시계열 데이터 처리에 특화    

- **키워드** : 시퀀스, 은닉 상태, LSTM, GRU

- **개념**
   - 시퀀스 데이터 : 크기가 고정되어 있지 않은 데이터
   - 은닉 상태 : 네트워크가 시간에 따라 어떤 정보를 기억할지 결정
      - 은닉 상태 계산 : 현재 은닉 상태는 현재 입력과 이전 은닉 상태에 기반하여 계산
      - 출력 계산 : 현재 은닉 상태에 기반하여 계산
   - LSTM : RNN의 장기 의존성 문제를 완화하기 위해 설계되어 입력/출력/망각 3개의 Gate를 통해 정보 흐름 조절
      - Forget Gate : LSTM의 첫 단계로 기존 정보 중 어떤 정보 버릴지 선택
      - Input Gate : 입력 데이터 중 어떤 정보를 다음 상태로 저장할지 결정
      - Output Gate : 다음 상태로 어떤 정보를 내보낼지 선택
   - GRU : LSTM보다 단순화한 구조로 Reset Gate와 Update Gate만을 가지고 성능 확보

### 딥러닝 역사

- 딥러닝 도래 : 1998년 LeNet-5(최초의 합성곱 신경망 구조) -> 2012년 AlexNet(딥러닝 시대 개시한 모델)
   - AlexNet : GPU 활용한 병렬 학습, ReLU 활성화 함수, 여러 계층 구조 등의 특징
- 딥러닝 발전 
   - 2014년 VGG(합성곱 계층과 풀링 계층만으로 이뤄진 구조와 일관된 필터 크기와 레이어 깊이 확장) / Sequence-to-Sequence(Seq2Seq, 시퀀스를 입력 받아 다른 시퀀스를 출력 -> 기계 변역 분야에서 사용되며 인코더-디코더 두 부분 구성) 
   - 2015년 GoogLeNet(복잡한 Inception 모듈을 사용해 파라미터 수를 줄이며 성능 향상, VGG와 다르게 다양한 크기의 레이어) / ResNet(기울기 소실 문제 해결을 위해 Residual Connection 도입) 
   - 2017년 Transformer(Self-Attention 매커니즘으로 시퀀스 데이터 처리, RNN과 LSTM 대체)
- 최신 딥러닝 
   - 2018년 BERT(언어 이해를 위한 양방향 Transformer 구조로 단어의 의미를 파악) / GPT(사전 훈련된 트랜스포머 모델로 디코더만을 사용) 
   - 2020년 Vision Transformer(NLP에 적용된 트랜스포머 구조를 이미지 인식에 활용) -> 2022년 ChatGPT(챗봇 전용 모델)

## ❓ 어려웠던 부분  
- CNN과 RNN을 실제로 적용하면 어떻게 할 수 있는지 궁금합니다
- LLM에 관심이 있어 Transformer 이론을 좀 더 쉽지만 확실하게 이해하고 싶습니다

## 🔗 참고 자료 (선택)  
- LLM : https://www.ibm.com/kr-ko/think/topics/large-language-models
- Transformer : https://blogs.nvidia.co.kr/blog/what-is-a-transformer-model/

**Comment by parkjunohjjang12** [comment]
안녕하세요 이가은 님! 올려주신 내용 잘 읽었습니다! 

CNN과 RNN에 대한 전반적인 내용을 잘 정리해주신 것 같습니다. 해당 내용에 관해 어려웠던 부분에서 질문해 주셨는데, 이에 대해 피드백 드리도록 하겠습니다!

---
## 1️⃣ 어려웠던 부분 피드백

**CNN과 RNN을 실제로 적용하면 어떻게 할 수 있는가?**  

- 질문 의도를 제가 100% 정확히 이해하진 못했지만, 보통 **CNN은 이미지/공간 패턴**, **RNN은 시계열/순차 패턴**에 많이 사용됩니다.

1. **CNN**
- 가장 흔한 **이미지 분류**부터 **Object Detection**, **Segmentation**까지 다양합니다. 이 외에도 **Super-Resolution(해상도 복원)**, **Denoising(노이즈 제거)** 등 응용이 매우 많습니다.
- 사용하는 **모델 구조**와 **출력 형식**은 태스크에 따라 달라집니다(예: 분류는 클래스 확률, 검출은 바운딩 박스+클래스, 세그멘테이션은 픽셀 단위 마스크 등).

2. **RNN**
- 입력 순서를 고려해 다음 출력을 예측하는 구조라 **Time Series**나 **NLP**에서 많이 사용됩니다(요즘 NLP는 Transformer가 주류지만, RNN 계열을 이해해두면 큰 도움이 됩니다).
- 마찬가지로 **태스크(분류/생성/예측)**에 따라 모델 구조와 출력이 달라집니다.

**Transformer 이론**  
- Transformer는 RNN의 장기 의존성·직렬 처리 한계를 보완하기 위해, 입력 토큰들에서 **Q, K, V**를 만들고 **Self-Attention**(예: `softmax(QK^T / sqrt(d_k)) V`)으로 **전역적으로 중요한 관계**를 학습하는 구조입니다.
- LLM뿐 아니라 **Vision(예: ViT)**에서도 폭넓게 쓰이며, 최근 SOTA 모델 상당수가 Transformer 변형을 기반으로 합니다.
- 더 자세한 설명은 아래 강의를 추천드립니다(저도 큰 도움을 받았습니다).
- 해당 강의 외에도 이전 강의에서 CNN이나 RNN 등 다양한 내용에 대해서는 올리셨으니, 강의를 들으시면서 이해가 안 되시는 부분이 있다면, 이전 강의 참고하셔서 수강하신느 것을 추천합니다!
  - 📄 [이준석 교수님 ML/DL Lec13: Transformer1](https://www.youtube.com/watch?v=NIFnKN2tWsE&list=PL0E_1UqNACXDTwuxUzCl5AeEjXBfWxCwc&index=14)  
  - 📄 [이준석 교수님 ML/DL Lec14: Transformer2](https://www.youtube.com/watch?v=LBqfClEnV2U&list=PL0E_1UqNACXDTwuxUzCl5AeEjXBfWxCwc&index=15)

---
LLM에 관심이 있으시다고 하셔서 Transformer 외에 최근 구글에서 발표한 논문 두 편도 공유드립니다.  
둘 다 긴 문맥(Long Context)에서 Transformer의 메모리/시간 복잡도(O(n²)) 한계를 보완하려는 메모리 모듈 계열입니다.

- **Titans**: Attention(단기·정밀) + **장기 메모리 모듈**을 결합해, 학습/추론 시 긴 과거 정보를 효율적으로 활용하도록 설계했습니다. (테스트 시점에서의 ‘메모라이즈’ 전략) :contentReference[oaicite:0]{index=0}
- **ATLAS**: 메모리의 **용량/업데이트/관리**를 더 최적화하여, 현재·과거 토큰을 함께 고려해 **테스트 시점에서 메모리를 학습적으로 최적화**하는 방식입니다. 논문에서는 이를 바탕으로 **DeepTransformers**라는 일반화 프레임을 제안합니다.

📄 원문  
- [Titans: *Learning to Memorize at Test Time*](https://arxiv.org/abs/2501.00663) 
- [ATLAS: *Learning to Optimally Memorize the Context at Test Time*](https://arxiv.org/abs/2505.23735)
감사합니다! 😀

**Comment by kkaeunii** [comment]
안녕하세요 멘토님!
답변 감사합니다! 보내주신 링크 잘 공부해보겠습니다 ㅎㅎ

첫번째 질문이 제가 다시 봐도 애매하네요 ㅎㅎ;
관련 실습 강의가 없어서, 실제 데이터셋에 어떻게 적용하는지 코드적으로 궁금했습니다.
이 부분은 좀 더 찾아보고 공부해보겠습니다! 감사합니다!

**Comment by parkjunohjjang12** [comment]
네! 궁금한 내용이 있으시다면 언제든지 편하게 질문주세요! :) 감사합니다~!
